include/seqan3/alphabet/composite/alphabet_variant.hpp

```diff
@@ -516,11 +535,17 @@ class alphabet_variant : public alphabet_base<alphabet_variant<alternative_types
         constexpr size_t table_size = 1 << (sizeof(char_type) * 8);
 
         std::array<rank_type, table_size> char_to_rank{};
+
+        for (size_t i = 0u; i < table_size; ++i)
+            char_to_rank[i] = seqan3::to_rank(assign_char_to(static_cast<char>(i), meta::front<alternatives>{}));
+
+        using index_t = std::make_unsigned_t<char_type>;
+        rank_type unknown = char_to_rank[std::min<index_t>(value_size, table_size - 1)];
+
         for (size_t i = 0u; i < rank_to_char.size(); ++i)
         {
-            using index_t = std::make_unsigned_t<char_type>;
             rank_type & old_entry = char_to_rank[static_cast<index_t>(rank_to_char[i])];
-            bool is_new_entry = rank_to_char[0] != rank_to_char[i] && old_entry == 0;
+            bool is_new_entry = rank_to_char[0] != rank_to_char[i] && (old_entry == unknown || value_size == table_size);
```

Comment:
Hm, I don't understand this fully...

Reply:
I first cach the `unknown` rank (e.g. that of `N` for dna5) to compare it here, similar as before where you compare to rank 0. BUT I determine the `unknown` rank by taking the rank **value_size** which is the first unknown rank. This could be undefined behaviour if `value_size == table_size` :( So I check for this when determining the `unknown` rank and I need to check here that I have a "valid unknown"..

any idea how to make this prettier?